src/plugins/intel_cpu/src/graph.cpp

```diff
@@ -1579,21 +1611,21 @@ bool Graph::InsertNode(EdgePtr edge, NodePtr node, bool initNode) {
     return InsertNode(edge->getParent(), edge->getChild(), node, iIndex, oIndex, initNode);
 }
 
+void Graph::InitNewNodes(const std::vector<NodePtr>& new_nodes) {
+    InitNodes(new_nodes);
+    InitDescriptors(new_nodes);
+    ResolveInplaceDirections(new_nodes);
+    InitOptimalPrimitiveDescriptors(new_nodes);
+}
+
 bool Graph::InsertNode(NodePtr parent, NodePtr child, NodePtr node, int parentPort, int childPort, bool initNode) {
     CreateEdge(parent, node, parentPort, 0);
     CreateEdge(node, child, 0, childPort);
-    node->updateConstantType();
-    AddNode(node);
 
-    if (initNode) {
-        node->getSupportedDescriptors();
-        node->initSupportedPrimitiveDescriptors();
-        node->filterSupportedPrimitiveDescriptors();
-        node->selectOptimalPrimitiveDescriptor();
-        resolveInPlaceDirection(node);
-        node->initOptimalPrimitiveDescriptor();
-    }
+    AddNode(node);
 
+    if (initNode)
+        InitNewNodes({node});
```

Comment:
According to your suggestion, I suppose the subgraph to be inserted has already been connected into the Graph by calling `Graph::CreateEdge` (recently introduced) before calling `Graph::InsertSubgraph`, so what `InsertSubgraph` is supposed to do is simply initialize the nodes contained in the subgraph which is exactly what `InitNewNodes` does.

If you mean `InsertSubgraph()` would also be responsible for connecting new nodes into subgraph and also connect the subgraph into main graph, I would say there is no simple way to provide the desired subgraph-topology information to `InsertSubgraph()`, I would rather call the flexible API `Graph::CreateEdge` to build the subgraph topology into the main graph, and then call `InitNewNodes()` to initialized newly created nodes in desired order.

Reply:
The main problem here is the fact that the graph has two main states, which are different from the transformation perspective: Uninitialized and Initialized. The `Graph::CreateEdge`, `Graph::AddNode` and other similar ops, allow us to easily manipulate the graph topology, but completely ignores the initialized state. Therefore, if we had made it correctly, they wouldn't have been allowed to be called to the Initialized (or partially initialized) graph, to avoid possible errors (as the current implementations of these calls invalidate the graph initialized state in general case). But, the initialized and unutilized states are not really separated and the graph object may be in both or even in-between. What your propose here is just a further step in this wrong direction. This method `InitNewNodes(const std::vector<NodePtr>& new_nodes)` disclose some internal complexity (as it become a public method) of the graph without binding it with all the other internal stuff properly, e.g. one may call initialization to nodes that are not a part of the graph, or to the nodes that are partially in the graph, so nothing prevents it from misuse. Also let's look at its purpose. You introduced this method, which has a decent impact on the graph interface, only to serve in a single transformation, which seems to be too much, and may be even resolved inside this transformation itself. On the other hand we already have `InsertNode` methods, which are beyond graph topology modification methods ( `Graph::CreateEdge`, `Graph::AddNode`) and consider the graph initialization specifics, though cumbersome (via the `initNode`) flag. So, to do it properly, we have to either extend the existing `InsertNode` action to a subgraph (when we insert a subgraph in-between two nodes) or somehow modify the graph topology modification methods ( `Graph::CreateEdge`, `Graph::AddNode` etc.) so that they respect the initialization state properly.